### PR TITLE
[SHIRO-552] Support base64 encoded salt in JdbcRealm

### PR DESCRIPTION
**Pull request by steinarb**
This fixes [SPAP-552](https://issues.apache.org/jira/browse/SHIRO-552)

The change has been tested in https://github.com/steinarb/authservice/ and worked as a drop in replacement for the existing [AuthserviceDbRealm ](https://github.com/steinarb/authservice/blob/master/authservice.web.security.dbrealm/src/main/java/no/priv/bang/authservice/web/security/dbrealm/AuthserviceDbRealm.java#L25)

Note that the change makes JdbcRealm expect salt to be base64 encoded by default.  It is possible to get the original behaviour back, by setting the [JdbcRealm saltIsBase64Encoded property](https://github.com/steinarb/shiro/blob/shiro-jdbcrealm-bas64-coded-salt/core/src/main/java/org/apache/shiro/realm/jdbc/JdbcRealm.java#L214) to false.

### CODE REVIEW

**Comment by bdemers** [comment]
@steinarb can you rebase your branch against master?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Shiro-pr/91/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Shiro-pr/92/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Shiro-pr/93/

**Comment by steinarb** [comment]
@bdemers I have rebased the branch against master and force-pushed (but I'm not really sure what the force push did to the pull request...?).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Shiro-pr/94/

**Comment by bdemers** [comment]
Thanks @steinarb! I just retargeted the PR against `master` (it was 1.4 previously)

File: core/src/main/java/org/apache/shiro/realm/jdbc/JdbcRealm.java [hunk 1]
```diff
@@ -247,7 +261,11 @@ protected AuthenticationInfo doGetAuthenticationInfo(AuthenticationToken token)
             info = new SimpleAuthenticationInfo(username, password.toCharArray(), getName());
             
             if (salt != null) {
-                info.setCredentialsSalt(ByteSource.Util.bytes(salt));
+            	if (saltStyle == SaltStyle.COLUMN && saltIsBase64Encoded) {
+                    info.setCredentialsSalt(ByteSource.Util.bytes(Base64.decode(salt)));
```
Comment (bdemers) [inline core/src/main/java/org/apache/shiro/realm/jdbc/JdbcRealm.java]: We could probably decode this once in `setSaltIsBase64Encoded()` and reduce the branch here

**Comment by bdemers** [comment]
@steinarb Looks good to me! can you sign an Apache CLA? https://www.apache.org/licenses/#contributor-license-agreements

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Shiro-pr/96/

**Comment by steinarb** [comment]
> @steinarb Looks good to me! can you sign an Apache CLA? https://www.apache.org/licenses/#contributor-license-agreements

@bdemers I have filled out the ICLA PDF and printed it to another PDF and signed the result with gpg and sent both the PDF and the asc to secretary@apache.org (I wasn't able to fill out anything in the signature field in the PDF but I assume the .asc takes care of that...?)

**Comment by bdemers** [comment]
@steinarb yup, you should be all set! 
Thanks again!!

**Comment by fpapon** [comment]
Hi @steinarb , any update about the ICLA?

**Comment by steinarb** [comment]
> Hi @steinarb , any update about the ICLA?

Hi @fpapon I understood from the comment by @bdemers above that it was OK...? 

In my INBOX there's received an automated response from secretary@apache.org, received 8 weeks ago, saying that "If you have been invited as a committer, please advise the (P)PMC that your ICLA has been filed."

**Comment by fpapon** [comment]
@steinarb ok, thanks!

**Comment by steinarb** [comment]
@bdemers  @fpapon  Should I resolve the conflict and re-push?

**Comment by fpapon** [comment]
@steinarb yes, thanks!

**Comment by steinarb** [comment]
@fpapon rebased, conflict resolved, built with "mvn clean install" with no errors, and branch force-pushed

**Pull request merged**